File: _BHCVRP_Python_Version/cujae_inf_citi_om/i_o/output/ExportResult.py
import json
import csv
import xml.etree.ElementTree as ET
from data.Problem import Problem
from data.ProblemType import ProblemType

class ExportResult:
    @staticmethod
    def to_txt(data, file_path):
        """Exporta los datos a un archivo TXT."""
        with open(file_path, "w") as file:
            file.write(data)

    @staticmethod
    def to_json(data, file_path):
        """Exporta los datos a un archivo JSON."""
        with open(file_path, "w") as file:
            json.dump(data, file, indent=4)

    @staticmethod
    def to_csv(data, file_path):
        """Exporta los datos a un archivo CSV."""
        with open(file_path, "w", newline="") as file:
            writer = csv.writer(file)
            writer.writerows(data)

    @staticmethod
    def to_xml(data, file_path):
        """
        Exporta los datos a un archivo XML.
        :param data: Diccionario con los datos a exportar.
        :param file_path: Ruta del archivo XML.
        """
        # Crear el elemento raíz
        root = ET.Element("resultado")

        # Agregar heurística, costo y tiempo
        ET.SubElement(root, "heuristica").text = data["heuristic_type"]
        ET.SubElement(root, "costo_total").text = str(data["total_cost"])
        ET.SubElement(root, "tiempo_ejecucion").text = str(data["execution_time"])

        # Agregar rutas
        rutas_element = ET.SubElement(root, "rutas")
        if Problem.get_problem().get_type_problem() == ProblemType.SBRP:
            for route in data["routes"]:
                ruta_element = ET.SubElement(rutas_element, "ruta", id=str(route["route_id"]))
                ET.SubElement(ruta_element, "paradas").text = ", ".join(map(str, route["bus_stops"]))
        else:
            for route in data["routes"]:
                ruta_element = ET.SubElement(rutas_element, "ruta", id=str(route["route_id"]))
                ET.SubElement(ruta_element, "clientes").text = ", ".join(map(str, route["customers"]))

        # Crear el árbol XML y guardar en un archivo
        tree = ET.ElementTree(root)
        tree.write(file_path, encoding="utf-8", xml_declaration=True)

File: BHCVRP_Java_Version/src/cujae/inf/citi/om/i_o/output/ExportResult.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package cujae.inf.citi.om.i_o.output;

import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import org.json.JSONObject;
import java.util.List;
import java.util.Map;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import org.w3c.dom.Document;
import org.w3c.dom.Element;

/**
 *
 * @author kmych
 */
public class ExportResult {

    public static void toTxt(String data, String filePath) throws IOException {
        try (FileWriter file = new FileWriter(filePath)) {
            file.write(data);
        }
    }

    public static void toJson(Map<String, Object> data, String filePath) throws IOException {
        JSONObject jsonObject = new JSONObject(data);
        try (FileWriter file = new FileWriter(filePath)) {
            file.write(jsonObject.toString(4)); // Indentación de 4 espacios
        }
    }

    public static void toCsv(ArrayList<ArrayList<String>> data, String filePath) throws IOException {
        try (FileWriter file = new FileWriter(filePath)) {
            for (List<String> row : data) {
                file.write(String.join(",", row) + "\n");
            }
        }
    }

    public static void toXml(Map<String, Object> data, String filePath) throws ParserConfigurationException, IOException, Exception {
        DocumentBuilderFactory docFactory = DocumentBuilderFactory.newInstance();
        DocumentBuilder docBuilder = docFactory.newDocumentBuilder();
        Document doc = docBuilder.newDocument();

        // Crear el elemento raíz
        Element rootElement = doc.createElement("resultado");
        doc.appendChild(rootElement);

        // Agregar heurística, costo y tiempo
        Element heuristicElement = doc.createElement("heuristica");
        heuristicElement.appendChild(doc.createTextNode((String) data.get("heuristic_type")));
        rootElement.appendChild(heuristicElement);

        Element costElement = doc.createElement("costo_total");
        costElement.appendChild(doc.createTextNode(String.valueOf(data.get("total_cost"))));
        rootElement.appendChild(costElement);

        Element timeElement = doc.createElement("tiempo_ejecucion");
        timeElement.appendChild(doc.createTextNode(String.valueOf(data.get("execution_time"))));
        rootElement.appendChild(timeElement);

        // Agregar rutas
        Element routesElement = doc.createElement("rutas");
        rootElement.appendChild(routesElement);

        List<Map<String, Object>> routes = (List<Map<String, Object>>) data.get("routes");
        for (Map<String, Object> route : routes) {
            Element routeElement = doc.createElement("ruta");
            routeElement.setAttribute("id", String.valueOf(route.get("route_id")));
            routesElement.appendChild(routeElement);

            Element customersElement = doc.createElement("clientes");
            customersElement.appendChild(doc.createTextNode(String.join(", ", (List<String>) route.get("customers"))));
            routeElement.appendChild(customersElement);
        }

        // Escribir el contenido en un archivo XML
        TransformerFactory transformerFactory = TransformerFactory.newInstance();
        Transformer transformer = transformerFactory.newTransformer();
        DOMSource source = new DOMSource(doc);
        StreamResult result = new StreamResult(new FileWriter(filePath));
        transformer.transform(source, result);
    }
}